### DISCUSSION TITLE
Correct way to enable App Attest provider for App Check in iOS

### DISCUSSION THREAD

**Discussion by xzilja** [invertase/react-native-firebase#6163] [Q&A]
I'm working on an app that supports iOS 14+ and it makes sense to use `App Attest` instead of default `DeviceCheck` provider in App Check.

From [RNFB docs](https://rnfirebase.io/reference/app-check#activate) I can see that recommended way to do this is via calling native methods in AppDelegate.

I wanted to clarify how exactly this should be done i.e. which imports and methods need to be called? To start off, here is my current AppDelegate initialisation of firebase:

```objc
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  // How to set FIRAppAttestProvider or FIRAppCheckDebugProviderFactory here and where to import these from?

  [FIRApp configure];
  
}
```

Ty, in advance

**Comment by mikehardy** [comment]
Hi there! You have chosen a difficult goal, if I am understanding the current state of firebase-ios-sdk + react-native-firebase AppCheck capabilities.

First, here's the best documentation on what is available:
https://firebase.google.com/docs/reference/ios/firebaseappcheck/api/reference/Classes

...of note is that they provide ProviderFactories for Debug tokens, and DeviceCheck tokens, but they do not provide an AppAttest provider factory, just the provider.

Which leads you to me note-from-the-past here:

https://github.com/invertase/react-native-firebase/blob/ef7895cd49eed3a4d103df624e52315250c68ab3/packages/app-check/ios/RNFBAppCheck/RNFBAppCheckModule.m#L110-L116

...where I am telling future-me or perhaps you "we need to write a custom factory for AppAttest" since apparently it's not provided and AppAttest isn't usable otherwise *if I am understanding correctly*

So the implementation steps are:
1. author something that implements https://firebase.google.com/docs/reference/ios/firebaseappcheck/api/reference/Protocols/FIRAppCheckProviderFactory but using the FIRAppCheckAppAttestProvider, patterned off these:
- https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRDeviceCheckProviderFactory.h
-  https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseAppCheck/Sources/DeviceCheckProvider/FIRDeviceCheckProviderFactory.m (I think?)

The code I linked above holding the "TODO" here in our module is how you would initialize it, prior to calling `[FIRApp configure]`

Note that if you were interested in doing this, my real dream is that the iOS AppCheck initialization was "delayed" just a little bit since we are not able - in interpreted languages - to delay the call to `[FIRApp configure]` that people put in `AppDelegate` as an initialization sequence thing, but I'd like for people to be able to do their fine configuration using javascript APIs, or at least firebase.json which could be package-specific and then the app-check ios module code could do things during it's init with an init provider of some sort.

Concretely what do I mean there?

It would be idea if react-native-firebase implemented an AppCheckTokenProviderFacadeFactory or similar, and used an init provider (that somehow runs before AppDelegate app launched code with `FIRApp configure` or just required manual integration, to install itself.

Now the dream is that within that facade, there was an implementation of a TokenProvider that was simply a facade we gave to the firebase-ios-sdk methods, and it delegated calls to the currently installed TokenProvider but allowed you to swap them via some APIs where javascript users could request (via an enum of strings or similar) the token provider factory they wanted, and it would include an AppAttestProviderFactory

That's the full dream anyway. The direct implementation of your desire is to implement AppAttestTokenProviderFactory returning already-built AppAttestTokenProvider from firebase-ios-sdk, and install that

**Reply by xzilja** [reply]
Gotcha, this is a bit over my head, but would factory like this make sense? (taken from firebase docs here https://firebase.google.com/docs/app-check/ios/app-attest-provider)

```objc
@interface YourAppCheckProviderFactory : NSObject <FIRAppCheckProviderFactory>
@end

@implementation YourAppCheckProviderFactory

- (nullable id<FIRAppCheckProvider>)createProviderWithApp:(nonnull FIRApp *)app {
  if (@available(iOS 14.0, *)) {
    return [[FIRAppAttestProvider alloc] initWithApp:app];
  } else {
    return [[FIRDeviceCheckProvider alloc] initWithApp:app];
  }
}

@end
```

And I assume you can then use it for ios implementation as it will fall back to `DeviceCheck`

**Reply by mikehardy** [reply]
I can't do a fallback like that since AppAttest actually requires some other configuration I think, on the xcproject + Apple web site ? (and we can't assume that's been done, or is desired)
But assuming you've done that....this seems like it would work? It isn't supposed to be complicated really - how it works - just really strict with how + when it initializes.

**Reply by xzilja** [reply]
Tried following after all of this, but I assume it is not correct as I am getting the error after

> Error: [appCheck/token-error] The operation couldn’t be completed. (com.firebase.appCheck error 0.)
NativeFirebaseError: [appCheck/token-error] The operation couldn’t be completed. (com.firebase.appCheck error 0.)

```objc
  [[FIRAppAttestProvider alloc] initWithApp:[FIRApp defaultApp]];
  
  [FIRApp configure];
```

You were right, seems like it'll be a hard task 😅 Will get some sleep on it and try again tomorrow

**Reply by mikehardy** [reply]
I think you have to make the factory and set the factory into AppCheck 🤔

**Reply by xzilja** [reply]
Wasn't able to figure it out in the end and went along with DeviceCheck, that one works really really nice 👌

**Reply by mikehardy** [reply]
Haha I don't blame you. We *will* have a better implementation of this at some point, when iOS14 as a minimum becomes a more reasonable stance for most apps and more look to adopt this

**Reply by xzilja** [reply]
I actually managed to implement it, writing the flow up now with few questions I want to ask you on cleanup and I think I can make a PR to docs explaining how to do it (as a work around)

**Reply by xzilja** [reply]
So step by step flow looked something like this for me (and this is how I'd lay it out in the docs in some sub section perhaps?)

1. Activate App Attest in firebase console by adding your apple `TeamID`

2. Create custom factory, I called it `AppAttestProviderFactory` and it is currently living in my `AppDelegate.m` file. This is actually my first question to you, how can I move it into a separate file? I tried to create `AppAttestProviderFactory.m` and import it in AppDelegate.m, but it failed with some error saying stuff alongside "double refference"
```objc
@interface AppAttestProviderFactory : NSObject<FIRAppCheckProviderFactory>
@end

@implementation AppAttestProviderFactory

- (nullable id<FIRAppCheckProvider>)createProviderWithApp:(nonnull FIRApp *)app {
  return [[FIRAppAttestProvider alloc] initWithApp:app];
}

@end
```

3. Create this factory and set up app check with it before your `[FIRApp configure]` call i.e.

```objc
AppAttestProviderFactory *providerFactory = [[AppAttestProviderFactory alloc] init];
[FIRAppCheck setAppCheckProviderFactory:providerFactory];

[FIRApp configure];
```

4. Added "App Attest" under app capability settings

5. Set it's value in `.entitlements` file to be "production" instead of development

---

And that's basically it. As a bonus I also implemented debug token via following additional steps

1. Added `-FIRDebugEnabled` flag under `Product` -> `Schema` -> `Edit Schema` -> `Run Debug` -> `Arguments`

2. Edited code above to be
```objc
  #if DEBUG
    FIRAppCheckDebugProviderFactory *providerFactory = [[FIRAppCheckDebugProviderFactory alloc] init];
  #else
    AppAttestProviderFactory *providerFactory = [[AppAttestProviderFactory alloc] init];
  #endif
  [FIRAppCheck setAppCheckProviderFactory:providerFactory];
  
  [FIRApp configure];
```

Although I'm sure I did it correctly, I am getting an error described here https://stackoverflow.com/questions/71677158/failing-to-get-debug-app-check-token-for-ios

**Reply by xzilja** [reply]
Let me know if any of this is interesting in terms of adding it to docs 👍 until it is exposed to js land

**Reply by mikehardy** [reply]
I think this is fantastic, and tested+working (almost no matter how...) is fine to just post up in case others need it (plus I can likely build off this later internally). 

Maybe...

- make a new page maybe app-attest.md in docs/app-attest/
- add it to the sidebar similar to others that have more than just usage page https://github.com/invertase/react-native-firebase/blob/ef7895cd49eed3a4d103df624e52315250c68ab3/docs/sidebar.yaml#L26-L28
- "break into" the current prev/next links to include it, likely best is to add app-attest as the next in docs/app-check/usage/index.md and then alter docs/app-distribution/usage/index.md to have app-attest.md as the new prev

Does that make sense? That's the basic way to slice in a new docs page, then I recommend just putting *exactly* what worked for you, so at least something is up and you know it worked. Iteration can always happen later if people want to extend+improve

**Reply by akeva001** [reply]
> So step by step flow looked something like this for me (and this is how I'd lay it out in the docs in some sub section perhaps?)
> 
> 1. Activate App Attest in firebase console by adding your apple `TeamID`
> 2. Create custom factory, I called it `AppAttestProviderFactory` and it is currently living in my `AppDelegate.m` file. This is actually my first question to you, how can I move it into a separate file? I tried to create `AppAttestProviderFactory.m` and import it in AppDelegate.m, but it failed with some error saying stuff alongside "double refference"
> 
> ```objc
> @interface AppAttestProviderFactory : NSObject<FIRAppCheckProviderFactory>
> @end
> 
> @implementation AppAttestProviderFactory
> 
> - (nullable id<FIRAppCheckProvider>)createProviderWithApp:(nonnull FIRApp *)app {
>   return [[FIRAppAttestProvider alloc] initWithApp:app];
> }
> 
> @end
> ```
> 
>  3. Create this factory and set up app check with it before your `[FIRApp configure]` call i.e.
> 
> ```objc
> AppAttestProviderFactory *providerFactory = [[AppAttestProviderFactory alloc] init];
> [FIRAppCheck setAppCheckProviderFactory:providerFactory];
> 
> [FIRApp configure];
> ```
> 
>  4. Added "App Attest" under app capability settings
> 5. Set it's value in `.entitlements` file to be "production" instead of development
> 
> And that's basically it. As a bonus I also implemented debug token via following additional steps
> 
> 1. Added `-FIRDebugEnabled` flag under `Product` -> `Schema` -> `Edit Schema` -> `Run Debug` -> `Arguments`
> 2. Edited code above to be
> 
> ```objc
>   #if DEBUG
>     FIRAppCheckDebugProviderFactory *providerFactory = [[FIRAppCheckDebugProviderFactory alloc] init];
>   #else
>     AppAttestProviderFactory *providerFactory = [[AppAttestProviderFactory alloc] init];
>   #endif
>   [FIRAppCheck setAppCheckProviderFactory:providerFactory];
>   
>   [FIRApp configure];
> ```
> 
> Although I'm sure I did it correctly, I am getting an error described here https://stackoverflow.com/questions/71677158/failing-to-get-debug-app-check-token-for-ios

Hi there! I am curious how you managed to get this working.  I followed your steps and still don't get verified requests to firebase. Would you be able to share your AppDelegate file? I added the provider factory the same as you did.

**Reply by mikehardy** [reply]
@akeva001 this is no longer necessary, AppAttest (and even "AppAttestWithDeviceCheckFallback" in case you need to support iOS 11-13, as AppAttest is iOS14+ only....) is supported officially in react-native-firebase v17.

**Reply by akeva001** [reply]
Thank you @mikehardy! I'll upgrade to v17.

**Reply by akeva001** [reply]
@mikehardy I have now upgraded to v17 but I am having trouble following the documentation for configuring the provider.

Is this the proper way to configure appCheck:
`rnfbProvider = firebase.appCheck().newReactNativeFirebaseAppCheckProvider();
rnfbProvider.configure({
  android: {
    provider: 'playIntegrity',
  },
  apple: {
    provider: 'appAttestWithDeviceCheckFallback',
  },
  web: {
    provider: 'reCaptchaV3',
    siteKey: 'unknown',
  },
});`

**Reply by mikehardy** [reply]
Yes, although if it t is a debug build you will likely want the debug provider in android and apple blocks so maybe a conditional in the config there

**Reply by akeva001** [reply]
@mikehardy I apologize, but I still can't get it working. I am testing on iOS and have added the Debug provider in the config.  Unfortunately, I am still not getting a verified request to the server.

I think I am not initializing app check properly within my firebase.js file?

Or it may be that I am unable to import newReactNativeFirebaseAppCheckProvider()?
Thank you again for all your help!

**Reply by akeva001** [reply]
Here is the error I am receiving when trying to use newReactNativeFirebaseAppCheckProvider():

TypeError: (0, _appCheck.newReactNativeFirebaseAppCheckProvider) is not a function.

**Reply by mikehardy** [reply]
If this is a javascript error (not a Typescript error), then that is unexpected and means that you are somehow executing javascript code that is not current, that is, you are not executing the current version of the app-check package, as that method exists and you can see it [here](https://github.com/invertase/react-native-firebase/blob/635d595389bfb224fc530681627fb2180fea6e26/packages/app-check/lib/index.js#L43), with example / testing usage [here](https://github.com/invertase/react-native-firebase/blob/635d595389bfb224fc530681627fb2180fea6e26/packages/app-check/e2e/appcheck.e2e.js#L22)

Why is that happening? A close inspection of your project is the only way to know, in the past I've seen old bundles dropped into apps from something like Expo OTA updater or Code Push. But it could be any of a number of things.

**Reply by akeva001** [reply]
Thank you of the response @mikehardy. Unfortunately I don't have over the air updates of any kind implemented. I opened the file where @firebase/appcheck (app-check-public.d.ts)  is defined and there is no export for newReactNativeFirebaseAppCheckProvider(). Could that be my problem?

**Reply by mikehardy** [reply]
Yes, the method is not defined in index.d.ts yet for app-check. It was an oversight. You can `// @ts-ignore` it for now. I'll fix it in a coming clean-up of the recent AppCheck work, see not-completed item 5 here --> #6886

**Reply by akeva001** [reply]
I see. Thank you @mikehardy

**Reply by akeva001** [reply]
Hey @mikehardy. So this is my current setup for appCheck. I believe everything is set up properly at this point. I put my debug token under the apple debugToken, but I am still no getting permission to access the server when I enforce AppCheck. Any ideas as to why this could be? Thanks in advance Mike!

<img width="488" alt="Screenshot 2023-02-10 at 10 50 40 AM" src="https://user-images.githubusercontent.com/57927148/218173945-d88cd7ab-68e7-4c27-ad05-155bd594e38c.png">

**Reply by mikehardy** [reply]
Unfortunately, as our documentation indicates, the debugToken param for apple platform is not working quite yet:

https://rnfirebase.io/app-check/usage#on-ios

> This method mainly consists of setting an environment variable in a way that it is visible to the Simulator as it starts up. That environment variable is automatically seen and used by the Firebase Debug provider on iOS.
> 
> In a future release, the debugToken parameter of the CustomProvider apple options should allow for more dynamic configuration, PRs to the Apple CustomProvider are welcome if you are motivated to implement this feature.


In the meantime setting the token by environment variable for the simulator works:

>  but the react-native-firebase testing app for e2e testing does successfully fetch App Check tokens via setting an environment variable

https://github.com/invertase/react-native-firebase/blob/635d595389bfb224fc530681627fb2180fea6e26/package.json#L45

**Reply by akeva001** [reply]
Thank you for all the help @mikehardy! Still unable to get it working even in a TestFlight build using "appAttestWithDeviceCheckFallback" as my provider. I honestly don't know what to do at this point. Been trying for months to get AppCheck working and still no progress. Everything else works great from Crashlytics to Analytics! I'll keep you posted if I figure it out in the future. Thanks again!

**Reply by akeva001** [reply]
Hey @mikehardy, so I was able to get this block of code to work and the error message is gone after I enter my debug token into the firebase console. 

<img width="505" alt="Screenshot 2023-02-14 at 8 45 29 PM" src="https://user-images.githubusercontent.com/57927148/218933582-ece4914a-3254-4634-94d7-f6d80c04263d.png">
But when I enforce app check from the firebase console, I receive a permissions error now. Any idea what I am missing?

<img width="1351" alt="Screenshot 2023-02-14 at 8 47 45 PM" src="https://user-images.githubusercontent.com/57927148/218933844-9b179cf1-aca7-45f6-88e9-dd849e5f8076.png">

**Reply by mikehardy** [reply]
I'm not sure - I can say that working on github with text in triple-backticks like so:

```
error message here
```

is a lot easier than graphics, that's really hard to read. It sounds like for some reason you are not getting valid app check tokens, I just can't say why. Configuring all the relevant items in the various consoles (Apple App config, entitlements, etc etc, Firebase console, etc) is non-trivial. It's also not clear what your execution environment is. I would not expect AppAttest or DeviceCheck to work on a Simulator. I wouldn't expect AppAttest to work on an iOS13 device, etc. It's not clear what you are trying, so it's not clear why it could fail.

**Reply by akeva001** [reply]
Hi @mikehardy, so I am testing on an iOS device running iOS 16.2. 

I am getting an appcheck token from ``` const tokenFB = await firebase.appCheck().getToken();```

When I enforce AppCheck on the FB console for both DB and Storage, I receive the following error on all my requests:
```FirebaseError: [code=permission-denied]: Missing or insufficient permissions.```

**Reply by mikehardy** [reply]
Okay, that seems like it should be a valid execution environment for the AppAttest provider (ios is 14+, you state you are on a real device). Why is your token not valid? I'm sorry, I just don't know. Some configuration issue like wrong bundle id name or wrong capitallization? Does debug provider or DeviceCheck provider work? (just curious). I'm sorry I can't say anything more specific. Seems like it *should* work.

**Comment by akeva001** [comment]
The debug provider does not work either. Do you think it could be because of the security rules I have set? I'll try to figure this out and keep you posted if I find a solution. Thanks @mikehardy!

**Reply by mikehardy** [reply]
You never want to run like this permanently, but you can always set an "allow everything" security rule as a *DANGEROUS* but *TEMPORARY* check just to see if it is security rules of course. Then immediately tighten the rules up again and debug them after verifying (or not) it was rules that were denying you

**Reply by akeva001** [reply]
Just tired with no luck. I just check my App Check API in the console and it seems like the requests are being made but I am not being permitted access to the DB. 

<img width="1108" alt="Screenshot 2023-02-15 at 1 30 19 PM" src="https://user-images.githubusercontent.com/57927148/219172327-1ebb9c5d-24f1-4f30-949e-485079cf1e11.png">

**Reply by akeva001** [reply]
@mikehardy I was able to fix my issue with appcheck. Now it all works! turns out the service file was the issue!

**Reply by libero-yoko** [reply]
@akeva001 
We are experiencing a very similar issue. We've noticed an excessive number of requests to ExchangeDebuToken, exceeding the 60 requests/minute quota.

Could you possibly clarify what the issue was in your service file? I'm assuming you're referring to the GoogleService-Info-[app].plist file. Thank you!

**Reply by aurelius0** [reply]
> @akeva001 We are experiencing a very similar issue. We've noticed an excessive number of requests to ExchangeDebuToken, exceeding the 60 requests/minute quota.
> 
> Could you possibly clarify what the issue was in your service file? I'm assuming you're referring to the GoogleService-Info-[app].plist file. Thank you!

Hi there @libero-yoko  - did you ever sort this out?  I have the same problem!  Thanks.
